pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/client.go

```diff
@@ -60,23 +62,20 @@ func NewClient() networkservice.NetworkServiceClient {
 	return &setKernelMacClient{}
 }
 
-func (s *setKernelMacClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*connection.Connection, error) {
-	conf := vppagent.Config(ctx)
-	conn, err := next.Client(ctx).Request(ctx, request, opts...)
-	if err != nil {
-		return nil, err
-	}
-	if mechanism := kernel.ToMechanism(request.GetConnection().GetMechanism()); mechanism != nil {
-		index := len(conf.GetLinuxConfig().GetInterfaces()) - 1
-		conf.GetLinuxConfig().GetInterfaces()[index+1].PhysAddress = conn.GetContext().GetEthernetContext().GetSrcMac()
+func (c *setKernelMacClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*connection.Connection, error) {
+	config := vppagent.Config(ctx)
+	if mechanism := kernel.ToMechanism(request.GetConnection().GetMechanism()); mechanism != nil && len(config.GetLinuxConfig().GetInterfaces()) > 0 {
+		current := len(config.LinuxConfig.Interfaces) - 1
+		config.LinuxConfig.Interfaces[current].PhysAddress = request.GetConnection().GetContext().GetEthernetContext().GetSrcMac()
 	}
-	return conn, nil
+	return next.Client(ctx).Request(ctx, request, opts...)
```

Comment:
You really have to call the next.Client here *before* you setup the config because the fully completed connection will come *back* from the return to that call, it doesn't come *in* on the Request for the client.